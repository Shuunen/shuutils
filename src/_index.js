'use strict'

import { getTimestamp, getTimestampMs } from './timestamps'

module.exports = { getTimestamp, getTimestampMs }