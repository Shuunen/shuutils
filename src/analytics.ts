/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types */
import { sleep } from './functions'
import { Logger } from './logger'

type AnalyticsInstance = {
  identify: (userId: string, additionalData?: Record<string, unknown>) => Promise<void>
  page: () => Promise<void>
  track: (event: string, additionalData?: Record<string, unknown>) => Promise<void>
}

/**
 * Generic analytics wrapper, can be used with any analytics third party library
 * @example const analytics = new Analytics()
 * @example analytics.setupInMemory('my-app')
 * @example analytics.track('invoice-download')
 * @example analytics.track('click', { category: '...', label: '...', ... })
 */
// eslint-disable-next-line no-restricted-syntax
export class Analytics {
  private $app = ''
  private $instance: AnalyticsInstance | undefined = undefined
  private readonly $logger = new Logger()
  public readonly pile: string[] = []

  /**
   * Identify a user
   * @param userId - The user, like '123456', 'super-admin', etc.
   * @param additionalData - Additional data to identify the user, like `{ email: '...', age: 35, ... }`
   */
  public identify(userId: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : identify failed, analytics not setup')
    this.$logger.info(`analytics : identify user "${userId}" on app "${this.$app}"`)
    void this.$instance.identify(userId, additionalData)
  }

  /**
   * Track a page view
   */
  public page() {
    if (this.$instance === undefined) throw new Error('analytics : track page failed, analytics not setup')
    this.$logger.debug(`analytics : track page for app "${this.$app}"`)
    void this.$instance.page()
  }

  /**
   * Setup analytics with in memory tracking, useful for tests
   * @param app - The name of the app, like 'mlm-react-front', 'cosmos-front', etc.
   */
  public setupInMemory(app: string) {
    this.$app = app
    // eslint-disable-next-line jsdoc/require-jsdoc
    const addToPile = async (item: string, data?: Record<string, unknown>) => {
      this.$logger.debug(`analytics : add "${item}" to pile`, data)
      await sleep(1)
      this.pile.push(item)
      this.$logger.debug(`analytics : pile for app "${app}"`, this.pile)
    }
    this.$instance = {
      // eslint-disable-next-line jsdoc/require-jsdoc
      identify: async (userId: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`identify:${userId}`, additionalData)
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      page: async () => {
        await addToPile('page')
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      track: async (event: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`track:${event}`, additionalData)
      },
    }
    this.$logger.info(`analytics : setup app "${app}" with in memory tracking`)
  }

  /**
   * Track an event
   * @param event - The event name like 'click', 'download-invoice', etc.
   * @param additionalData - Additional data to track the event, like `{ category: '...', label: '...', ... }`
   */
  public track(event: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : track event failed, analytics not setup')
    this.$logger.debug(`analytics : track event "${event}" for app "${this.$app}"`)
    void this.$instance.track(event, additionalData)
  }
}
