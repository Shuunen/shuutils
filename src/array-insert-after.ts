import { nbAfter, nbNoIndex } from './constants'
import { clone } from './object-clone'

/**
 * Add a value to an array if it doesn't already exist
 * @param array the array to add the value to
 * @param item the value to add
 * @param value the value to add
 * @returns the array with the value added (if it didn't already exist)
 */
export function insertValueAfterItem<T> (array: T[], item: T, value: T) {
  const index = array.indexOf(item)
  if (index === nbNoIndex) return array
  const arrayCopy = clone(array)
  const start = Math.min(index + nbAfter, arrayCopy.length)
  arrayCopy.splice(start, 0, value)
  return arrayCopy
}
