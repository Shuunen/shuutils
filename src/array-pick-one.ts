/**
 * Return an item from an array
 * @param items like : ["great", "place", "pine"]
 * @returns item like : "pine"
 */
export function pickOne<Type>(items: readonly Type[]) {
  if (items.length === 0) throw new Error('Array is empty')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return items[Math.floor(Math.random() * items.length)] as Type
}
