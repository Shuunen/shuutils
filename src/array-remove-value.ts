import { clone } from './object-clone'

/**
 * Remove a value from an array
 * @param array the array to remove the value from
 * @param value the value to remove
 * @returns a copy of the array without the value
 */
export function removeValueFromArray<Type>(array: readonly Type[], value: Type) {
  if (!array.includes(value)) return array
  const arrayCopy = clone(array)
  arrayCopy.splice(arrayCopy.indexOf(value), 1)
  return arrayCopy
}
