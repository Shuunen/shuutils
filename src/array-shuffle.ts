import { Nb } from './constants'
import { clone } from './object-clone'

/**
 * Randomize an array using the Fisher-Yates algorithm
 * source : https://dev.to/codebubb/how-to-shuffle-an-array-in-javascript-2ikj
 * @param input the array to shuffle
 * @returns a shuffled clone of the input array
 */
export function shuffleArray<T> (input: T[]) {
  const array = clone(input)
  // eslint-disable-next-line for-direction
  for (let indexA = array.length - 1; indexA > 0; indexA += Number(Nb.Previous)) {
    const indexB = Math.floor(Math.random() * (indexA + 1))
    const temporary = array[indexA]
    array[indexA] = array[indexB] as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
    array[indexB] = temporary as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
  }
  return array
}
