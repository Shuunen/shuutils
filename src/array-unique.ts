/**
 * Return an array with unique values
 * @param items the array to check/clean/reduce
 * @returns the array with unique values
 */
export function arrayUnique<Type>(items: readonly Type[]) {
  return Array.from(new Set(items))
}
