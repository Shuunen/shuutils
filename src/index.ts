export * from './array-insert-after'
export * from './array-pick-one'
export * from './array-remove-value'
export * from './array-shuffle'
export * from './array-unique'
export * from './browser-clipboard'
export * from './browser-fill-like-human'
export * from './browser-inject-styles'
export * from './browser-page-change'
export * from './browser-scout'
export * from './browser-toast'
export * from './colors'
export * from './constants'
export * from './date-relative'
export * from './dates'
export * from './debounce'
export * from './dom'
export * from './environment'
export * from './events'
export * from './expect-type'
export * from './functions'
export * from './logger'
export * from './memoize'
export * from './number-random'
export * from './number-ranged-score'
export * from './number-round'
export * from './object-cap-nesting'
export * from './object-clone'
export * from './object-equal'
export * from './object-flatten'
export * from './object-serializer'
export * from './object-sort'
export * from './objects'
export * from './state'
export * from './storage'
export * from './string-remove-accents'
export * from './strings'
export * from './testing-align-for-snap'
export * from './throttle'
export * from './timestamps'
export * from './types'
