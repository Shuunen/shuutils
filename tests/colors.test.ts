import { test } from 'uvu'
import { ok } from 'uvu/assert'
import { hexToHsl, red, yellow } from '../src'
import { check } from './utils'

test('colored strings are still strings :)', function () {
  console.log(yellow(':D'))
  ok(red('plop'))
})

check('color red short : hex => hsl', hexToHsl('#f00'), { h: 0, s: 100, l: 50 })
check('color red long : hex => hsl', hexToHsl('#ff0000'), { h: 0, s: 100, l: 50 })
check('color rebeccapurple : hex => hsl', hexToHsl('#663399'), { h: 270, s: 50, l: 40 })

test.run()
