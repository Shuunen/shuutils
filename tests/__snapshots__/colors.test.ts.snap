// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`color bgBlack 1`] = `"[40mtest[49m"`;

exports[`color bgBlue 1`] = `"[44mtest[49m"`;

exports[`color bgCyan 1`] = `"[46mtest[49m"`;

exports[`color bgGray 1`] = `"[100mtest[49m"`;

exports[`color bgGreen 1`] = `"[42mtest[49m"`;

exports[`color bgMagenta 1`] = `"[45mtest[49m"`;

exports[`color bgRed 1`] = `"[41mtest[49m"`;

exports[`color bgWhite 1`] = `"[47mtest[49m"`;

exports[`color bgYellow 1`] = `"[43mtest[49m"`;

exports[`color black 1`] = `"[30mtest[39m"`;

exports[`color blue 1`] = `"[34mtest[39m"`;

exports[`color bold 1`] = `"[1mtest[22m"`;

exports[`color cyan 1`] = `"[36mtest[39m"`;

exports[`color dim 1`] = `"[2mtest[22m"`;

exports[`color gray 1`] = `"[90mtest[39m"`;

exports[`color green 1`] = `"[32mtest[39m"`;

exports[`color inverse 1`] = `"[7mtest[27m"`;

exports[`color italic 1`] = `"[3mtest[23m"`;

exports[`color magenta 1`] = `"[35mtest[39m"`;

exports[`color overline 1`] = `"[53mtest[55m"`;

exports[`color red 1`] = `"[31mtest[39m"`;

exports[`color reset 1`] = `"[0mtest[0m"`;

exports[`color strikeThrough 1`] = `"[9mtest[29m"`;

exports[`color underline 1`] = `"[4mtest[24m"`;

exports[`color white 1`] = `"[37mtest[39m"`;

exports[`color yellow 1`] = `"[33mtest[39m"`;
